Fix: Prevent <|nocaptions|> tokens in BatchedInferencePipeline

mmichelli opened the pull request:
https://github.com/SYSTRAN/faster-whisper/issues/1319#issuecomment-3153705309


- Add nocaptions component tokens [1771, 496, 9799] to suppress_tokens list
- Add segment filtering to remove any remaining <|nocaptions|> segments
- Resolves issue where BatchedInferencePipeline would generate malformed special tokens during periods of silence or low-confidence transcription
- Includes comprehensive tests to verify the fix

The issue occurred because while bracket tokens ('<', '|', '>') were already suppressed, the content tokens ('no', 'ca', 'ptions') were not, leading to partial token generation that formed complete <|nocaptions|> tags in the output.

Files changed:
- faster_whisper/transcribe.py: Core fix implementation
- test_nocaptions_comprehensive.py: Comprehensive test suite
- tests/test_nocaptions_fix.py: Unit tests

MahmoudAshraf97 commented on PR_DESCRIPTION.md:
Remove this

MahmoudAshraf97 commented on test_nocaptions_comprehensive.py:
remove this

MahmoudAshraf97 commented on tests/test_nocaptions_fix.py:
remove this

MahmoudAshraf97 reviewed (changes requested):
Hello, `<|nocaptions|>` is a single token not a multiple token segment
what needs to be done here is add a `no_speech` attribute to the tokenizer which should be equal to `model.hf_tokenizer.token_to_id("<|nocaptions|>")` and add this to the suppressed tokens

mmichelli commented:
I've updated the PR to implement the single token approach as suggested by @MahmoudAshraf97.

**Changes made:**
- Added `no_speech` property to Tokenizer class using `tokenizer.token_to_id("<|nocaptions|>")`
- Added null checks to handle models without this token
- Updated `get_suppressed_tokens` to conditionally include the `no_speech` token
- Updated tests with correct token ID expectations

**Testing:**
- All tokeniser tests pass
- Works with different model variants (tiny, large-v3, etc.)
- No breaking changes

mmichelli commented:
I thought it was possible that the `<|nocaptions|>` were missing.

For example:
https://huggingface.co/NbAiLab/nb-whisper-large/blob/main/tokenizer.json

```python
from faster_whisper import WhisperModel

model = WhisperModel("NbAiLab/nb-whisper-large", device="cuda")
nocaptions_token_id = model.hf_tokenizer.token_to_id("<|nocaptions|>")
print(f"<|nocaptions|> token ID: {nocaptions_token_id}")
```

**<|nocaptions|> token ID: None**

RafalCer commented:
Hey, @mmichelli I ran into the same issue with NB whisper, i.e. constant `<|nocaptions|>` on segments with background noise or multiple people talking at the same time. 

As you noted in your previous comment, the token (`<|nocaptions|>`) is not available in the tokenizer of the model. You can print out the token indices to see that whenever NB Whisper's output is `<|nocaptions|>` , the model actually outputs a sequence of tokens to construct it, as you can see in the example below:

```
Subsegment tokens: [27, 91, 1771, 496, 9799, 91, 29]
Subsegment text: <|nocaptions|>
```

I tried adding `<|nocaptions|>` to vocabulary, but ultimately solved the issue for me was adding  `<` id, which is 27, to `suppress_ids_begin` in `config.json`. To the best of my understanding, this way all outputs starting with `<` get suppressed and NB whisper outputs its second best guess for the sequence. From my quick evaluation, the second best guess for what would otherwise be `<|nocaptions|>` is almost always correct. Plus, this way I have way less gaps in the transcript.

MahmoudAshraf97 commented:
This is a problem with the finetune itself, it's not a case we should handle in this PR, and users of this model will be able to pass any extra tokens they want to suppress to `transcribe` function
